Documentation/guides/D8andR8.md

```diff
@@ -0,0 +1,270 @@
+  --output <file>          # Output result in <file>.
+                           # <file> must be an existing directory or a zip file.
+  --lib <file>             # Add <file> as a library resource.
+  --min-api                # Minimum Android API level compatibility.
+  --pg-conf <file>         # Proguard configuration <file>.
+  --pg-map-output <file>   # Output the resulting name and line mapping to <file>.
+  --no-tree-shaking        # Force disable tree shaking of unreachable classes.
+  --no-minification        # Force disable minification of names.
+  --no-desugaring          # Force disable desugaring.
+  --main-dex-rules <file>  # Proguard keep rules for classes to place in the
+                           # primary dex file.
+  --main-dex-list <file>   # List of classes to place in the primary dex file.
+  --main-dex-list-output <file>  # Output the full main-dex list in <file>.
+  --version                # Print the version of r8.
+  --help                   # Print this message.
+```
+
+# What did Xamarin.Android do *before* D8/R8?
+
+In other words, what is currently happening *before* we introduce D8/R8 support?
+
+1. The [Javac](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/Javac.cs)
+   MSBuild task compiles `*.java` files to a `classes.zip` file.
+2. The [Desugar](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/Desugar.cs)
+   MSBuild task "desugars" using `desugar_deploy.jar` if
+   `$(AndroidEnableDesugar)` is `True`.
+3. The [Proguard](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/Proguard.cs)
+   MSBuild task shrinks the compiled Java code if
+   `$(AndroidEnableProguard)` is `True`. Developers may also supply
+   custom proguard configuration files via `ProguardConfiguration`
+   build items.
+4. The [CreateMultiDexMainDexClassList](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/CreateMultiDexMainDexClassList.cs)
+   MSBuild task runs `proguard` to generate a final, combined
+   `multidex.keep` file if `$(AndroidEnableMultiDex)` is `True`.
+   Developers can also supply custom `multidex.keep` files via
+   `MultiDexMainDexList` build items.
+5. The [CompileToDalvik](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs)
+   MSBuild task runs `dx.jar` to generate a final `classes.dex` file
+   in `$(IntermediateOutputPath)android\bin`. If `multidex` is
+   enabled, a `classes2.dex` (and potentially more) are also generated
+   in this location.
+
+# What will this process look like with D8 / R8?
+
+Two new MSBuild tasks named `R8` and `D8` will be created.
```

Comment:
Tense: when this PR is merged and becomes "proper" documentation, `R8` and `D8` *will exist*, not "will be created".  Thus:

> The `<D8/>` and `<R8/>` MSBuild tasks are used to invoke the corresponding Android SDK tools.